From eeafe30279bf1d73aa0a0928cbc3ddeae29be6b5 Mon Sep 17 00:00:00 2001
From: Romulo Leite <romulo.leite@windriver.com>
Date: Fri, 5 Sep 2025 16:33:56 -0300
Subject: [PATCH] Add retry to hostname reading by neutron agents

This patch adds a retry for stx-openstack neutron agents'
initialization. It was observed in different occasions
on standard deployments with multiple hosts that a race
condition causes neutron agents pods to fail to get the host name.

Signed-off-by: Romulo Leite <romulo.leite@windriver.com>
---
 .../bin/_neutron-dhcp-agent-init.sh.tpl       | 23 ++++++++++++++++++-
 .../bin/_neutron-l3-agent-init.sh.tpl         | 23 ++++++++++++++++++-
 .../_neutron-linuxbridge-agent-init.sh.tpl    | 23 ++++++++++++++++++-
 .../bin/_neutron-metadata-agent-init.sh.tpl   | 23 ++++++++++++++++++-
 .../_neutron-openvswitch-agent-init.sh.tpl    | 23 ++++++++++++++++++-
 .../bin/_neutron-ovn-vpn-agent-init.sh.tpl    | 23 ++++++++++++++++++-
 .../bin/_neutron-sriov-agent-init.sh.tpl      | 23 ++++++++++++++++++-
 7 files changed, 154 insertions(+), 7 deletions(-)

diff --git a/neutron/templates/bin/_neutron-dhcp-agent-init.sh.tpl b/neutron/templates/bin/_neutron-dhcp-agent-init.sh.tpl
index 3df3315bc..112d2d028 100644
--- a/neutron/templates/bin/_neutron-dhcp-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-dhcp-agent-init.sh.tpl
@@ -17,9 +17,30 @@ limitations under the License.
 set -ex
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME" ] || [ $RETRY_COUNT -eq 5 ]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
diff --git a/neutron/templates/bin/_neutron-l3-agent-init.sh.tpl b/neutron/templates/bin/_neutron-l3-agent-init.sh.tpl
index b9b93b28a..f70aa9448 100644
--- a/neutron/templates/bin/_neutron-l3-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-l3-agent-init.sh.tpl
@@ -17,9 +17,30 @@ limitations under the License.
 set -ex
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME" ] || [ $RETRY_COUNT -eq 5 ]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
diff --git a/neutron/templates/bin/_neutron-linuxbridge-agent-init.sh.tpl b/neutron/templates/bin/_neutron-linuxbridge-agent-init.sh.tpl
index f46d9a3ea..f6b7d9ab1 100644
--- a/neutron/templates/bin/_neutron-linuxbridge-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-linuxbridge-agent-init.sh.tpl
@@ -58,9 +58,30 @@ local_ip = "${LOCAL_IP}"
 EOF
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME" ] || [ $RETRY_COUNT -eq 5 ]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
diff --git a/neutron/templates/bin/_neutron-metadata-agent-init.sh.tpl b/neutron/templates/bin/_neutron-metadata-agent-init.sh.tpl
index 1cb25e30e..9f21dfc29 100644
--- a/neutron/templates/bin/_neutron-metadata-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-metadata-agent-init.sh.tpl
@@ -22,9 +22,30 @@ chown ${NEUTRON_USER_UID} /run/openvswitch/db.sock
 {{- end }}
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME" ] || [ $RETRY_COUNT -eq 5 ]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
diff --git a/neutron/templates/bin/_neutron-openvswitch-agent-init.sh.tpl b/neutron/templates/bin/_neutron-openvswitch-agent-init.sh.tpl
index 8e2c25e7c..c3a9ba7f0 100644
--- a/neutron/templates/bin/_neutron-openvswitch-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-openvswitch-agent-init.sh.tpl
@@ -567,9 +567,30 @@ EOF
 fi
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME"] || [$RETRY_COUNT -eq 5]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME = $(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
diff --git a/neutron/templates/bin/_neutron-ovn-vpn-agent-init.sh.tpl b/neutron/templates/bin/_neutron-ovn-vpn-agent-init.sh.tpl
index 5b6ce43e1..341462287 100644
--- a/neutron/templates/bin/_neutron-ovn-vpn-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-ovn-vpn-agent-init.sh.tpl
@@ -19,9 +19,30 @@ set -ex
 chown ${NEUTRON_USER_UID} /var/lib/neutron/openstack-helm
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME" ] || [ $RETRY_COUNT -eq 5 ]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
diff --git a/neutron/templates/bin/_neutron-sriov-agent-init.sh.tpl b/neutron/templates/bin/_neutron-sriov-agent-init.sh.tpl
index d98cfe8f2..0ea7f513d 100644
--- a/neutron/templates/bin/_neutron-sriov-agent-init.sh.tpl
+++ b/neutron/templates/bin/_neutron-sriov-agent-init.sh.tpl
@@ -96,10 +96,31 @@ ethtool --set-priv-flags ${NIC_FIRST_PORT} vf-true-promisc-support ${promisc_mod
 {{- end }}
 
 {{- if and ( empty .Values.conf.neutron.DEFAULT.host ) ( .Values.pod.use_fqdn.neutron_agent ) }}
+
+HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+
+# If the neutron agent can't get the node host name, try again
+
+if [ -z $HOST_NAME ]; then
+   echo "Could not get host name from node, retrying it "
+   RETRY_COUNT=0
+   until [ -n "$HOST_NAME" ] || [ $RETRY_COUNT -eq 5 ]; do
+      sleep 2
+      RETRY_COUNT=$((RETRY_COUNT+1))
+      HOST_NAME=$(hostname --fqdn 2>/dev/null || true)
+   done
+fi
+
+# If host is still empty, exit with error
+if [ -z $HOST_NAME ]; then
+   echo "Error: Could not get host name from node"
+   exit 1
+fi
+
 mkdir -p /tmp/pod-shared
 tee > /tmp/pod-shared/neutron-agent.ini << EOF
 [DEFAULT]
-host = $(hostname --fqdn)
+host = ${HOST_NAME}
 EOF
 {{- end }}
 
-- 
2.34.1

